tempesta_fw/sched/tfw_sched_rr.c

```diff
@@ -90,43 +90,67 @@ tfw_sched_rr_add_conn(TfwSrvGroup *sg, TfwServer *srv, TfwConnection *conn)
 
 	srv_cl = &sl->srvs[s];
 	for (c = 0; c < srv_cl->conn_n; ++c)
-		if (srv_cl->conns[c] == conn) {
+		if (srv_cl->conns[c] == srv_conn) {
 			TFW_WARN("sched_rr: Try to add existing connection,"
-				 " srv=%d conn=%d\n", s, c);
+				 " srv=%zu conn=%zu\n", s, c);
 			return;
 		}
-	srv_cl->conns[c] = conn;
+	srv_cl->conns[c] = srv_conn;
 	++srv_cl->conn_n;
 	BUG_ON(srv_cl->conn_n > TFW_SRV_MAX_CONN);
 }
 
 /**
- * On each subsequent call the function returns the next server in the group.
- * Parallel connections to the same server are also rotated in the
- * round-robin manner.
+ * On each subsequent call the function returns the next server in the
+ * group. Parallel connections to the same server are also rotated in
+ * the round-robin manner.
+ *
  * Dead connections and servers w/o live connections are skipped.
+ * Initially, connections with non-idempotent requests are also skipped
+ * in attempt to increase throughput. However, if all live connections
+ * contain a non-idempotent request, then re-run the algorithm and get
+ * the first live connection they way it is usually done.
+ *
+ * RR scheduler must be the fastest scheduler. Also, it's essential
+ * to maintain strict round-robin fashion of getting the next server.
+ * Usually the optimistic approach gives the fastest solution: we are
+ * optimistic in that there are not many non-idempotent requests, and
+ * there are available server connections.
  */
-static TfwConnection *
+static TfwSrvConn *
 tfw_sched_rr_get_srv_conn(TfwMsg *msg, TfwSrvGroup *sg)
 {
-	int c, s, i;
-	TfwConnection *conn;
+	size_t c, s;
+	unsigned long idxval;
+	int skipnip = 1, nipconn = 0;
 	TfwRrSrvList *sl = sg->sched_data;
 	TfwRrSrv *srv_cl;
+	TfwSrvConn *srv_conn;
 
 	BUG_ON(!sl);
-
+rerun:
 	for (s = 0; s < sl->srv_n; ++s) {
-		i = atomic64_inc_return(&sl->rr_counter) % sl->srv_n;
-		srv_cl = &sl->srvs[i];
+		idxval = atomic64_inc_return(&sl->rr_counter);
```

Comment:
Well, I'd say that `i` is usually anything beginning with "i", like "index", "iterator", "idxval" and so on. Just a short name when the code and meaning of the variable is clear. If you feel much more comfortable with `idxval` here, I wouldn't mind.